clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerProtocol.java

```diff
@@ -35,287 +35,160 @@
-            new Field(USER_DATA_KEY_NAME, Type.NULLABLE_BYTES));
-
-    public static final Schema ASSIGNMENT_V1 = new Schema(
-        new Field(TOPIC_PARTITIONS_KEY_NAME, new ArrayOf(TOPIC_ASSIGNMENT_V0)),
-        new Field(USER_DATA_KEY_NAME, Type.NULLABLE_BYTES));
-
-    public static Short deserializeVersion(ByteBuffer buffer) {
-        Struct header = CONSUMER_PROTOCOL_HEADER_SCHEMA.read(buffer);
-        return header.getShort(VERSION_KEY_NAME);
+    static {
+        // Safety check to ensure that both parts of the consumer protocol remain in sync.
+        if (ConsumerProtocolSubscriptionData.LOWEST_SUPPORTED_VERSION
+                != ConsumerProtocolAssignmentData.LOWEST_SUPPORTED_VERSION)
+            throw new IllegalStateException("Subscription and Assignment schemas must have the " +
+                "same lowest version");
+
+        if (ConsumerProtocolSubscriptionData.HIGHEST_SUPPORTED_VERSION
+                != ConsumerProtocolAssignmentData.HIGHEST_SUPPORTED_VERSION)
+            throw new IllegalStateException("Subscription and Assignment schemas must have the " +
+                "same highest version");
     }
 
-    public static ByteBuffer serializeSubscriptionV0(Subscription subscription) {
-        Struct struct = new Struct(SUBSCRIPTION_V0);
-        struct.set(USER_DATA_KEY_NAME, subscription.userData());
-        struct.set(TOPICS_KEY_NAME, subscription.topics().toArray());
+    public static short deserializeVersion(final ByteBuffer buffer) {
+        try {
+            return buffer.getShort();
+        } catch (BufferUnderflowException e) {
+            throw new SchemaException("Buffer underflow while parsing consumer protocol's header", e);
+        }
+    }
 
-        ByteBuffer buffer = ByteBuffer.allocate(CONSUMER_PROTOCOL_HEADER_V0.sizeOf() + SUBSCRIPTION_V0.sizeOf(struct));
-        CONSUMER_PROTOCOL_HEADER_V0.writeTo(buffer);
-        SUBSCRIPTION_V0.write(buffer, struct);
-        buffer.flip();
-        return buffer;
+    public static ByteBuffer serializeSubscription(final Subscription subscription) {
+        return serializeSubscription(subscription, ConsumerProtocolSubscriptionData.HIGHEST_SUPPORTED_VERSION);
     }
 
-    public static ByteBuffer serializeSubscriptionV1(Subscription subscription) {
-        Struct struct = new Struct(SUBSCRIPTION_V1);
-        struct.set(USER_DATA_KEY_NAME, subscription.userData());
-        struct.set(TOPICS_KEY_NAME, subscription.topics().toArray());
-        List<Struct> topicAssignments = new ArrayList<>();
+    public static ByteBuffer serializeSubscription(final Subscription subscription, short version) {
+        version = checkSubscriptionVersion(version);
+
+        ConsumerProtocolSubscriptionData data = new ConsumerProtocolSubscriptionData();
+        data.setTopics(subscription.topics());
+        if (subscription.userData() != null)
+            data.setUserData(subscription.userData().array());
```

Comment:
This might not be safe. If we use the "zero-copy" flag as suggested below, we can just duplicate the ByteBuffer instead.